instrumentation/kotlinx-coroutines/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/kotlinxcoroutines/instrumentationannotations/AnnotationInstrumentationHelper.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.kotlinxcoroutines.instrumentationannotations;
+
+import static io.opentelemetry.javaagent.instrumentation.kotlinxcoroutines.instrumentationannotations.AnnotationSingletons.instrumenter;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanKind;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.instrumentation.api.util.VirtualField;
+import kotlin.coroutines.Continuation;
+import kotlin.coroutines.intrinsics.IntrinsicsKt;
+
+public final class AnnotationInstrumentationHelper {
+
+  private static final VirtualField<Continuation<?>, Context> contextField =
+      VirtualField.find(Continuation.class, Context.class);
+
+  public static MethodRequest createMethodRequest(
+      Class<?> declaringClass, String methodName, String withSpanValue, String spanKindString) {
+    SpanKind spanKind = SpanKind.INTERNAL;
+    if (spanKindString != null) {
+      try {
+        spanKind = SpanKind.valueOf(spanKindString);
+      } catch (IllegalArgumentException exception) {
+        // ignore
+      }
+    }
+
+    return MethodRequest.create(declaringClass, methodName, withSpanValue, spanKind);
+  }
+
+  public static Context enterCoroutine(
+      int label, Continuation<?> continuation, MethodRequest request) {
+    // label 0 means that coroutine is started, any other label means that coroutine is resumed
```

Comment:
Do you know of any documentation/kotlin design doc that'd explain the `label` value? I couldn't find any, unfortunately.

Reply:
https://www.youtube.com/watch?v=YrrUCSi72E8 was useful if I remember correctly (around 7:30 there is something about labels). Basically it should work so that when coroutine is suspended it returns the label where it should be resumed and on resume it is called with the same label.

Reply:
Thanks! I'll watch that